src/Build/Graph/ProjectInterpretation.cs

```diff
@@ -0,0 +1,387 @@
+                        String.Format(
+                            CultureInfo.InvariantCulture,
+                            ResourceUtilities.GetResourceString(
+                                "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                            projectReference.EvaluatedInclude,
+                            requesterInstance.FullPath));
+                }
+
+                var projectReferenceFullPath = projectReference.GetMetadataValue(FullPathMetadataName);
+
+                var referenceGlobalProperties = GetGlobalPropertiesForItem(projectReference, requesterInstance.GlobalPropertiesDictionary, globalPropertiesModifiers);
+
+                var referenceConfig = new ConfigurationMetadata(projectReferenceFullPath, referenceGlobalProperties);
+
+                yield return referenceConfig;
+            }
+        }
+
+        private static string GetInnerBuildPropertyValue(ProjectInstance project)
+        {
+            return project.GetPropertyValue(GetInnerBuildPropertyName(project));
+        }
+
+        private static string GetInnerBuildPropertyName(ProjectInstance project)
+        {
+            return project.GetPropertyValue(PropertyNames.InnerBuildProperty);
+        }
+
+        private static string GetInnerBuildPropertyValues(ProjectInstance project)
+        {
+            return project.GetPropertyValue(project.GetPropertyValue(PropertyNames.InnerBuildPropertyValues));
+        }
+
+        internal static ProjectType GetProjectType(ProjectInstance project)
+        {
+            var isOuterBuild = String.IsNullOrWhiteSpace(GetInnerBuildPropertyValue(project)) && !String.IsNullOrWhiteSpace(GetInnerBuildPropertyValues(project));
+            var isInnerBuild = !String.IsNullOrWhiteSpace(GetInnerBuildPropertyValue(project));
+
+            ErrorUtilities.VerifyThrow(!(isOuterBuild && isInnerBuild), $"A project cannot be an outer and inner build at the same time: ${project.FullPath}");
+
+            return isOuterBuild
+                ? ProjectType.OuterBuild
+                : isInnerBuild
+                    ? ProjectType.InnerBuild
+                    : ProjectType.NonMultitargeting;
+        }
+
+        public static void PostProcess(ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode> allNodes)
+        {
+            foreach (var nodeKvp in allNodes)
+            {
+                var outerBuild = nodeKvp.Value;
+
+                if (GetProjectType(outerBuild.ProjectInstance) == ProjectType.OuterBuild && outerBuild.ReferencingProjects.Count != 0)
+                {
+                    foreach (var innerBuild in outerBuild.ProjectReferences)
+                    {
+                        foreach (var referencingProject in outerBuild.ReferencingProjects)
+                        {
+                            referencingProject.AddProjectReference(innerBuild);
+                        }
+                    }
+
+                    outerBuild.RemoveReferences();
```

Comment:
Ah, my confusion reduces:

This removes links from outer->inner builds of any build that is not a graph root, because the what-targets-should-we-run-on-a-ProjectInstance code has already decided not to run `Build` (really DispatchToInnerBuilds) on the outer build.

The referencing project, then, is given results for

* Outer build: GetTargetFrameworks
* Inner builds: Actual build

And can call its negotiation code followed by an inner build using those results.

This makes sense but has a surprising (to me) property: adding a new graph root that just calls the prior graph root changes the graph in such a way that **the new graph is not a subgraph of the old one**. Is that ok? I think it probably is, but deserves documentation.

This might have negative implications for graph caching. Is the edge removal required?

Reply:
The edge removal is there to have the graph be a more truthful representation of the runtime graph. The outer -> inner edges for non root outer builds do not appear in reality. Because of this, anyone who reasons about the graph needs to know about three little idiosyncrasies:
1. that outer -> inner edges are not there in reality. So one cannot reason about a single node in isolation, one always needs to check for this case.
2. out of all the outer -> inner edges, only one manifests in reality.
3. the edge to the outer build node is a true edge

With the edge removal, the number of idiosyncrasies get reduced to only one, a variation of 2: When referencing a multitargeting node, the referencing node considers all TFs and gets to pick one. The graph contains all the edges, similar to symbolic execution, but at runtime only one path is taken.

You are right, the structural difference between outer builds as roots and outer builds as non roots destroys the nice complete-structural-sharing property for graph caching. Here's some thoughts on that:
- This issue exists with vanilla msbuild, it's not an additional artificial static graph consequence
- Even though it's not 100% sharing, I think there's still significant structural sharing. Given a graph G, we can decompose it into the graph of roots, Gr, and the graph of non-roots, Gnr. When a new node N is added to depend on G via an edge inside Gr, Gr cannot be reused, but Gnr can. The trick will be to keep track of Gr and Gnr.

Options
1. change the graph to enable full structural sharing of subgraphs. The compromise here is that the new graph structure might depart from the actual runtime graph
2. keep the graph in tune with the runtime graph, but make graph caching more complex.

For now I'd just go with 2. I'll update the spec to account for this though.

Reply:
Sounds like a good plan.

Reply:
Hopefully in hindsight it still will have sounded like a good plain :)
cc @AndyGerlicher @dannyvv for visibility.